// @roblox-ts/luau-ast@1.0.11/out/LuauAST/impl/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntaxKind = void 0;
var SyntaxKind;
(function (SyntaxKind) {
    SyntaxKind[SyntaxKind["Identifier"] = 0] = "Identifier";
    SyntaxKind[SyntaxKind["TemporaryIdentifier"] = 1] = "TemporaryIdentifier";
    SyntaxKind[SyntaxKind["ComputedIndexExpression"] = 2] = "ComputedIndexExpression";
    SyntaxKind[SyntaxKind["PropertyAccessExpression"] = 3] = "PropertyAccessExpression";
    SyntaxKind[SyntaxKind["CallExpression"] = 4] = "CallExpression";
    SyntaxKind[SyntaxKind["MethodCallExpression"] = 5] = "MethodCallExpression";
    SyntaxKind[SyntaxKind["ParenthesizedExpression"] = 6] = "ParenthesizedExpression";
    SyntaxKind[SyntaxKind["None"] = 7] = "None";
    SyntaxKind[SyntaxKind["NilLiteral"] = 8] = "NilLiteral";
    SyntaxKind[SyntaxKind["FalseLiteral"] = 9] = "FalseLiteral";
    SyntaxKind[SyntaxKind["TrueLiteral"] = 10] = "TrueLiteral";
    SyntaxKind[SyntaxKind["NumberLiteral"] = 11] = "NumberLiteral";
    SyntaxKind[SyntaxKind["StringLiteral"] = 12] = "StringLiteral";
    SyntaxKind[SyntaxKind["VarArgsLiteral"] = 13] = "VarArgsLiteral";
    SyntaxKind[SyntaxKind["FunctionExpression"] = 14] = "FunctionExpression";
    SyntaxKind[SyntaxKind["BinaryExpression"] = 15] = "BinaryExpression";
    SyntaxKind[SyntaxKind["UnaryExpression"] = 16] = "UnaryExpression";
    SyntaxKind[SyntaxKind["IfExpression"] = 17] = "IfExpression";
    SyntaxKind[SyntaxKind["InterpolatedString"] = 18] = "InterpolatedString";
    SyntaxKind[SyntaxKind["Array"] = 19] = "Array";
    SyntaxKind[SyntaxKind["Map"] = 20] = "Map";
    SyntaxKind[SyntaxKind["Set"] = 21] = "Set";
    SyntaxKind[SyntaxKind["MixedTable"] = 22] = "MixedTable";
    SyntaxKind[SyntaxKind["Assignment"] = 23] = "Assignment";
    SyntaxKind[SyntaxKind["BreakStatement"] = 24] = "BreakStatement";
    SyntaxKind[SyntaxKind["CallStatement"] = 25] = "CallStatement";
    SyntaxKind[SyntaxKind["ContinueStatement"] = 26] = "ContinueStatement";
    SyntaxKind[SyntaxKind["DoStatement"] = 27] = "DoStatement";
    SyntaxKind[SyntaxKind["WhileStatement"] = 28] = "WhileStatement";
    SyntaxKind[SyntaxKind["RepeatStatement"] = 29] = "RepeatStatement";
    SyntaxKind[SyntaxKind["IfStatement"] = 30] = "IfStatement";
    SyntaxKind[SyntaxKind["NumericForStatement"] = 31] = "NumericForStatement";
    SyntaxKind[SyntaxKind["ForStatement"] = 32] = "ForStatement";
    SyntaxKind[SyntaxKind["FunctionDeclaration"] = 33] = "FunctionDeclaration";
    SyntaxKind[SyntaxKind["MethodDeclaration"] = 34] = "MethodDeclaration";
    SyntaxKind[SyntaxKind["VariableDeclaration"] = 35] = "VariableDeclaration";
    SyntaxKind[SyntaxKind["ReturnStatement"] = 36] = "ReturnStatement";
    SyntaxKind[SyntaxKind["Comment"] = 37] = "Comment";
    SyntaxKind[SyntaxKind["MapField"] = 38] = "MapField";
    SyntaxKind[SyntaxKind["InterpolatedStringPart"] = 39] = "InterpolatedStringPart";
    SyntaxKind[SyntaxKind["FirstIndexableExpression"] = 0] = "FirstIndexableExpression";
    SyntaxKind[SyntaxKind["LastIndexableExpression"] = 6] = "LastIndexableExpression";
    SyntaxKind[SyntaxKind["FirstExpression"] = 0] = "FirstExpression";
    SyntaxKind[SyntaxKind["LastExpression"] = 22] = "LastExpression";
    SyntaxKind[SyntaxKind["FirstStatement"] = 23] = "FirstStatement";
    SyntaxKind[SyntaxKind["LastStatement"] = 37] = "LastStatement";
    SyntaxKind[SyntaxKind["FirstField"] = 38] = "FirstField";
    SyntaxKind[SyntaxKind["LastField"] = 39] = "LastField";
})(SyntaxKind || (exports.SyntaxKind = SyntaxKind = {}));
